kube-apiserver crashes when trying to patch TPR data

**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

No.

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

* patch panic

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.1.535+73fb978181dcb2", GitCommit:"73fb978181dcb2e1aa54d09cbd0c48826f63d39b", GitTreeState:"clean", BuildDate:"2017-04-17T22:11:49Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.1.535+73fb978181dcb2", GitCommit:"73fb978181dcb2e1aa54d09cbd0c48826f63d39b", GitTreeState:"clean", BuildDate:"2017-04-17T22:09:04Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: GCE
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:


**What happened**:

* Try to do strategic merge patch on a TPR data object with `kubectl patch`.
* kube-apiserver crashes with a nil pointer dereference on [this line](https://github.com/enisoc/kubernetes/blob/c5e2f7c229d8194386de97df7ff6c1b0f71be4c7/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go#L610-L610):
  ```go
  currentVersionedObject, err := unsafeConvertor.ConvertToVersion(currentObject, kind.GroupVersion())
  ```

kube-apiserver.log:
```
runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:479
/usr/local/go/src/runtime/panic.go:458
/usr/local/go/src/runtime/panic.go:62
/usr/local/go/src/runtime/sigpanic_unix.go:24
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:610
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/update.go:183
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:396
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:484
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:291
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:474
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:746
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:1070
/usr/local/go/src/runtime/asm_amd64.s:2086
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xdae9c8]

goroutine 13356 [running]:
panic(0x30287c0, 0xc420014030)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0xc4245f7f28, 0x1, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x126
panic(0x30287c0, 0xc420014030)
	/usr/local/go/src/runtime/panic.go:458 +0x243
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.patchResource.func1(0x7f344d369c60, 0xc424739560, 0x0, 0x0, 0x5a6f040, 0xc4235b19e0, 0x5c2c820, 0x7f344d1d38a0, 0x37, 0x3)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:610 +0x1748
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest.(*defaultUpdatedObjectInfo).UpdatedObject(0xc4251e2040, 0x7f344d369c60, 0xc424739560, 0x5a6f040, 0xc4235b19e0, 0xf0, 0xf0, 0x31f8180, 0xc4245f77c8)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/update.go:183 +0xa1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Update.func1(0x5a6f040, 0xc4235b19e0, 0x0, 0x37e, 0x0, 0x0, 0x0, 0x0, 0xc4251e2168)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:396 +0xde
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*store).updateState(0xc420bd2cb0, 0xc4251e22c0, 0xc420b822a0, 0x5a6f040, 0xc4235b19e0, 0x0, 0x0, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:484 +0x5b
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*store).GuaranteedUpdate(0xc420bd2cb0, 0x7f344d51a550, 0xc424739560, 0xc4246b34a0, 0x57, 0x5a6f040, 0xc4235b18c0, 0x1, 0xc423238018, 0xc420b822a0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:291 +0x492
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Update(0xc421391800, 0x7f344d369c60, 0xc424739560, 0xc422516628, 0x12, 0x5a86ec0, 0xc4251e2040, 0x0, 0x0, 0x5a96b00, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:474 +0x3be
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.patchResource.func3(0xc400000018, 0x42b75c0, 0xc4208a7758, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:746 +0xbb
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.finishRequest.func1(0xc4246b32c0, 0xc422517270, 0xc4246b3260, 0xc4246b3200)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:1070 +0x99
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.finishRequest
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:1075 +0xd9
```

**What you expected to happen**:

I don't expect strategic merge patch to work on TPR data, but kube-apiserver should not crash.

**How to reproduce it** (as minimally and precisely as possible):

```console
$ cat tpr-crontab.yaml 
apiVersion: extensions/v1beta1
kind: ThirdPartyResource
metadata:
  name: cron-tab.stable.example.com
description: "A specification of a Pod to run on a cron style schedule"
versions:
- name: v1

$ kubectl create -f tpr-crontab.yaml 
thirdpartyresource "cron-tab.stable.example.com" created

$ cat my-crontab.yaml 
apiVersion: "stable.example.com/v1"
kind: CronTab
metadata:
  name: my-new-cron-object
cronSpec: "* * * * /5"
image: old-image

$ kubectl create -f my-crontab.yaml 
crontab "my-new-cron-object" created

$ kubectl patch crontab my-new-cron-object -p '{"image":"new-image"}' -v=9
[...]
I0417 15:55:57.279664   26411 request.go:991] Request Body: {"image":"new-image"}
I0417 15:55:57.279741   26411 round_trippers.go:398] curl -k -v -XPATCH  -H "Content-Type: application/strategic-merge-patch+json" -H "User-Agent: kc17/v1.7.0 (linux/amd64) kubernetes/73fb978" -H "Authorization: Bearer eW8qijm0xQAgNpUCPEMSs5G6xThXDRKP" -H "Accept: application/json" https://104.155.173.241/apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object
I0417 15:55:57.327149   26411 round_trippers.go:417] PATCH https://104.155.173.241/apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object  in 47 milliseconds
I0417 15:55:57.327182   26411 round_trippers.go:423] Response Headers:
F0417 15:55:57.327237   26411 helpers.go:119] error: unexpected EOF
```

**Anything else we need to know**:

My guess is that it's because `UnsafeConvertor` is `nil` for TPR data resources:

https://github.com/enisoc/kubernetes/blob/464db160b410b22dba53af6d11fa097f0fa4fd6b/pkg/master/thirdparty/thirdparty.go#L314-L337

--- a/pkg/master/thirdparty/thirdparty.go
+++ b/pkg/master/thirdparty/thirdparty.go
@@ -315,11 +315,12 @@ func (m *ThirdPartyResourceServer) thirdpartyapi(group, kind, version, pluralRes
 		Root:         apiRoot,
 		GroupVersion: externalVersion,
 
-		Creater:   thirdpartyresourcedata.NewObjectCreator(group, version, api.Scheme),
-		Convertor: api.Scheme,
-		Copier:    api.Scheme,
-		Defaulter: api.Scheme,
-		Typer:     api.Scheme,
+		Creater:         thirdpartyresourcedata.NewObjectCreator(group, version, api.Scheme),
+		Convertor:       api.Scheme,
+		Copier:          api.Scheme,
+		Defaulter:       api.Scheme,
+		Typer:           api.Scheme,
+		UnsafeConvertor: api.Scheme,
 
 		Mapper:                 thirdpartyresourcedata.NewMapper(api.Registry.GroupOrDie(extensions.GroupName).RESTMapper, kind, version, group),
 		Linker:                 api.Registry.GroupOrDie(extensions.GroupName).SelfLinker,

